File: frontend/src/types/rbac.ts
/**
 * RBAC（Role-Based Access Control）関連の型定義
 */

/**
 * ロールの種類
 */
export enum RoleType {
  ADMIN = 'ADMIN',
  PROJECT_LEADER = 'PROJECT_LEADER',
  MEMBER = 'MEMBER'
}

/**
 * ロール
 */
export interface Role {
  id: number;
  name: RoleType;
  description: string;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * パーミッション
 */
export interface Permission {
  id: number;
  name: string;
  resource: string;
  action: string;
  description: string;
  created_at: string;
  updated_at: string;
}

/**
 * ユーザーロール
 */
export interface UserRole {
  id: number;
  user_id: number;
  role_id: number;
  project_id: number | null;
  role: Role;
  created_at: string;
  updated_at: string;
}

/**
 * 権限チェック用のユーティリティ型
 */
export interface PermissionCheck {
  hasRole: (role: RoleType, projectId?: number) => boolean;
  hasPermission: (permission: string, projectId?: number) => boolean;
  canAccessProject: (projectId: number) => boolean;
  canManageProject: (projectId: number) => boolean;
  isAdmin: () => boolean;
  isProjectLeader: () => boolean;
}

/**
 * ロール階層の定義
 */
export const ROLE_HIERARCHY = {
  [RoleType.ADMIN]: [RoleType.ADMIN, RoleType.PROJECT_LEADER, RoleType.MEMBER],
  [RoleType.PROJECT_LEADER]: [RoleType.PROJECT_LEADER, RoleType.MEMBER],
  [RoleType.MEMBER]: [RoleType.MEMBER]
};

/**
 * パーミッション定数
 */
export const PERMISSIONS = {
  // ユーザー管理
  USERS_READ: 'users.read',
  USERS_WRITE: 'users.write',
  USERS_DELETE: 'users.delete',
  
  // プロジェクト管理
  PROJECTS_READ: 'projects.read',
  PROJECTS_WRITE: 'projects.write',
  PROJECTS_DELETE: 'projects.delete',
  PROJECTS_MANAGE: 'projects.manage',
  
  // メトリクス
  METRICS_READ: 'metrics.read',
  METRICS_EXPORT: 'metrics.export',
  
  // システム管理
  SYSTEM_ADMIN: 'system.admin'
} as const;

/**
 * ロールごとのデフォルトパーミッション
 */
export const ROLE_PERMISSIONS = {
  [RoleType.ADMIN]: Object.values(PERMISSIONS),
  [RoleType.PROJECT_LEADER]: [
    PERMISSIONS.PROJECTS_READ,
    PERMISSIONS.PROJECTS_WRITE,
    PERMISSIONS.PROJECTS_MANAGE,
    PERMISSIONS.METRICS_READ,
    PERMISSIONS.METRICS_EXPORT
  ],
  [RoleType.MEMBER]: [
    PERMISSIONS.PROJECTS_READ,
    PERMISSIONS.METRICS_READ
  ]
};

File: backend/app/models/rbac.py
from sqlalchemy import Column, String, Boolean, Integer, ForeignKey, UniqueConstraint, Table
from sqlalchemy.orm import relationship
from app.db.base_class import BaseModel

# Association table for many-to-many relationship between roles and permissions
role_permissions = Table(
    'role_permissions',
    BaseModel.metadata,
    Column('role_id', Integer, ForeignKey('team_insight.roles.id'), primary_key=True),
    Column('permission_id', Integer, ForeignKey('team_insight.permissions.id'), primary_key=True),
    schema='team_insight'
)

class Role(BaseModel):
    """
    ロールモデル
    
    システム内の役割を定義します。
    例: ADMIN, PROJECT_LEADER, MEMBER
    """
    __tablename__ = "roles"
    __table_args__ = {'schema': 'team_insight'}

    name = Column(String(50), unique=True, nullable=False, index=True)
    description = Column(String(255))
    is_system = Column(Boolean, default=False)  # システムロール（削除不可）
    
    # Relationships
    permissions = relationship("Permission", secondary=role_permissions, back_populates="roles")
    user_roles = relationship("UserRole", back_populates="role", cascade="all, delete-orphan")

class Permission(BaseModel):
    """
    パーミッションモデル
    
    システム内の権限を定義します。
    例: users.read, users.write, projects.manage
    """
    __tablename__ = "permissions"
    __table_args__ = {'schema': 'team_insight'}

    name = Column(String(100), unique=True, nullable=False, index=True)
    resource = Column(String(50), nullable=False)  # リソース名（例: users, projects）
    action = Column(String(50), nullable=False)    # アクション名（例: read, write, delete）
    description = Column(String(255))
    
    # Relationships
    roles = relationship("Role", secondary=role_permissions, back_populates="permissions")

class UserRole(BaseModel):
    """
    ユーザーロール関連モデル
    
    ユーザーとロールの関連を管理します。
    プロジェクト単位でのロール割り当てもサポートします。
    """
    __tablename__ = "user_roles"
    __table_args__ = (
        UniqueConstraint('user_id', 'role_id', 'project_id', name='_user_role_project_uc'),
        {'schema': 'team_insight'}
    )

    user_id = Column(Integer, ForeignKey('team_insight.users.id'), nullable=False)
    role_id = Column(Integer, ForeignKey('team_insight.roles.id'), nullable=False)
    project_id = Column(Integer, nullable=True)  # NULL = グローバルロール
    
    # Relationships
    user = relationship("User", back_populates="user_roles")
    role = relationship("Role", back_populates="user_roles")